tensorrec/loss_graphs.py

```diff
@@ -54,59 +63,50 @@ def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
-
-        tf_overlap_distribution = tf.contrib.distributions.Normal(loc=(tf_neg_mean - tf_pos_mean),
-                                                                  scale=tf.sqrt(tf_neg_var + tf_pos_var))
-
-        loss = 1.0 - tf_overlap_distribution.cdf(0.0)
-        return loss
-
-
 class WMRBLossGraph(AbstractLossGraph):
     """
     Approximation of http://ceur-ws.org/Vol-1905/recsys2017_poster3.pdf
     Interactions can be any positive values, but magnitude is ignored. Negative interactions are also ignored.
     """
     is_dense = True
+    is_sample_based = True
 
-    def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
+    def loss_graph(self, tf_prediction, tf_interactions, tf_sample_predictions, **kwargs):
+
+        # WMRB expects bounded predictions
+        tanh_prediction = tf.nn.sigmoid(tf_prediction)
+        tanh_sample_prediction = tf.nn.sigmoid(tf_sample_predictions)
 
+        return self.weighted_margin_rank_batch(tf_prediction=tanh_prediction,
+                                               tf_interactions=tf_interactions,
+                                               tf_sample_predictions=tanh_sample_prediction)
+
+    @classmethod
+    def weighted_margin_rank_batch(cls, tf_prediction, tf_interactions, tf_sample_predictions):
         positive_interaction_mask = tf.greater(tf_interactions.values, 0.0)
         positive_interaction_indices = tf.boolean_mask(tf_interactions.indices,
                                                        positive_interaction_mask)
+
+        # [ n_positive_interactions ]
         positive_predictions = tf.gather_nd(tf_prediction, indices=positive_interaction_indices)
 
         n_items = tf.cast(tf.shape(tf_prediction)[1], dtype=tf.float32)
-        predictions_sum_per_user = tf.reduce_sum(tf_prediction, axis=1)
-        mapped_predictions_sum_per_user = tf.gather(params=predictions_sum_per_user,
-                                                    indices=tf.transpose(positive_interaction_indices)[0])
+        n_sampled_items = tf.cast(tf.shape(tf_sample_predictions)[1], dtype=tf.float32)
 
-        # TODO smart irrelevant item indicator -- using n_items is an approximation for sparse interactions
-        irrelevant_item_indicator = n_items # noqa
+        # [ n_positive_interactions, n_sampled_items ]
+        mapped_predictions_sample_per_interaction = tf.gather(params=tf_sample_predictions,
+                                                              indices=tf.transpose(positive_interaction_indices)[0])
 
-        sampled_margin_rank = (n_items - (n_items * positive_predictions)
-                               + mapped_predictions_sum_per_user + irrelevant_item_indicator)
+        # [ n_positive_interactions, n_sampled_items ]
+        summation_term = tf.abs(1.0
+                                - tf.expand_dims(positive_predictions, axis=1)
+                                + mapped_predictions_sample_per_interaction)
```

Comment:
What about the indicator? assuming this is rolled into `positive_predictions`, `mapped_predictions_sample_per_interaction`?

Reply:
If we (naiively) assume that the indicator is 1 for all sampled items, and since its multiplied in, I just didn't write it in here. I'll add a TODO noting that the irrelevant item indicator isn't in place.